### DISCUSSION TITLE
Visual C++ project and properties

### DISCUSSION THREAD

**Discussion by Agrael1** [VsixCommunity/Community.VisualStudio.Toolkit#350] [Q&A] [answered]
Hello, I'm working on extension for C++ include management. Right now I have to use Com interfaces which have little to no documentation. Will there be a port of VCProject compiler and properties?

**Comment by madskristensen** [comment]
I've never worked with C++ from a extensibility perspective, so I'm not sure what is involved in doing so

**Reply by Agrael1** [reply]
The most common things queried in our project is compiler options. That requires several queries: DTE->Project as VCProject->IVCRulePropertyStorage2 | VCCLCompilerTool which are quire cumbersome. As of now I could not find any converters from VS.Solution.GetActiveProjectAsync() to VCProject neither are there any guides on how to get project property page values.

**Reply by Agrael1** [reply]
Also, is there any command to save al current files, without querying them?

**Reply by kaheimri** [reply, marked as answer]
> The most common things queried in our project is compiler options. That requires several queries: DTE->Project as VCProject->IVCRulePropertyStorage2 | VCCLCompilerTool which are quire cumbersome. As of now I could not find any converters from VS.Solution.GetActiveProjectAsync() to VCProject neither are there any guides on how to get project property page values.

You can convert a Community.VisualStudio.Toolkit.Project to a VCProject like this:

```c#
private const int ExtObject = (int)__VSHPROPID.VSHPROPID_ExtObject;
public static async Task<VCProject?> ConvertToVcProjectAsync(Community.VisualStudio.Toolkit.Project? project)
{
    if (project == null)
        return null;

    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();

    project.GetItemInfo(out var hierarchy, out _, out _);
    if (hierarchy.GetProperty(VSConstants.VSITEMID_ROOT, ExtObject, out var obj) != 0)
        return null;

    return (obj as EnvDTE.Project)?.Object as VCProject;
}
```
